Several API cleanups + additions

MJDSys opened the pull request:
This PR has several cleanups to the API to better align with with I see other rust crates do.  I also fill out the rest of libsystemd function interface.

I tried to break out each change into its own commit, so if any are unwanted I'll remove them.  The only one I'd really want to see kept is in 	accf112, as its a correctness change.  8731e61 exposed an issue I fixed as well, put I can remove the libsystemd-sys change if you just want the bug fix.

Let me know if you have any questions about anything!

jmesmon commented:
Things I like:

 - using enums instead of c_void
 - filling out the remaining ffi functions
 - completing modularization of ffi
 - pointer fixes

Things I don't:

 - removing $crate from sd_try! is likely not correct in all situations. We might want to consider just not exporting it (in which case it would then be OK to remove $crate, but I'm not sure why we wouldn't just keep it).
 - The re-exports of ffi types are a hedge against libraries using rust-systemd having a different version of libc than rust-systemd, and remove the question of whether to use the std::os::raw or libc:: types. Essentially, these types are already part of our public API (for libsystemd-sys) because the functions their accept them. Even if rust doesn't currently force us to export all the types we use in out public API (if they are provided by other crates), in this case I think it makes sense to do so. There are a few outstanding RFCs/issues regarding cargos crate selection that make this a good idea (ie: we can't tell it that out public API uses a crate with version X).

 - For similar reasons, I'm not really happy about restricting the version requirements. Technically semver doesn't provide any promises below version 1. In practice, some crate providers try to provide some promises. But I don't think the pay off we get for restricting versions (avoiding breakage) is really worth it here, because they don't really avoid breakage totally and it means there are a whole bunch of extra version numbers I have to remember to update.

MJDSys commented:
Regarding the $crate removal, that was an over eager regex.  Since that $crate removal came from one of the now removed commits, that's fixed.  I'm not sure if the macro should be public, but that's an issue for another PR.

Regarding your other dislikes, I figured those may not be wanted.  I just figured I'd submit it all, and if you didn't like it I'd make it easy to remove.  I've updated the PR to with the changes, and I updated the login.rs/event.rs files to use types exposed by the crate to match the other submodules.

jmesmon commented:
Great, thanks. Yep, no problem submitting those extra changes. It might make sense for me to add some comments explaining the choices around those areas.
